File: problems/3-sum/cpp/main.cpp
#include <algorithm>
#include <vector>

class Solution {
public:
  std::vector<std::vector<int>> threeSum(std::vector<int> &nums) {
    std::sort(nums.begin(), nums.end());
    std::vector<std::vector<int>> results;
    for (auto i = 0; i < nums.size(); ++i) {
      if (i != 0 && nums[i - 1] == nums[i]) {
        continue;
      }
      auto left = i + 1;
      auto right = nums.size() - 1;
      while (left < right) {
        const auto sum = nums[i] + nums[left] + nums[right];
        if (sum < 0) {
          ++left;
        } else if (sum > 0) {
          --right;
        } else {
          results.push_back({nums[i], nums[left], nums[right]});
          while (++left < right && nums[left - 1] == nums[left])
            ;
        }
      }
    }
    return results;
  }
};

File: problems/3-sum/python/main.py
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        results = []
        for i, n in enumerate(nums):
            if i != 0 and nums[i - 1] == n:
                continue
            left = i + 1
            right = len(nums) - 1
            while left < right:
                sum = n + nums[left] + nums[right]
                if sum < 0:
                    left += 1
                elif sum > 0:
                    right -= 1
                else:
                    results.append([n, nums[left], nums[right]])
                    left += 1
                    while left < right and nums[left - 1] == nums[left]:
                        left += 1
        return results
